pkg/sql/tenant.go

```diff
@@ -618,3 +629,36 @@ func TestingUpdateTenantRecord(
 ) error {
 	return updateTenantRecord(ctx, execCfg, txn, info)
 }
+
+// RenameTenant implements the tree.TenantOperator interface.
+func (p *planner) RenameTenant(ctx context.Context, tenID uint64, name string) error {
+	if err := p.RequireAdminRole(ctx, "create tenant"); err != nil {
+		return err
+	}
+
+	if !p.EvalContext().Settings.Version.IsActive(ctx, clusterversion.TenantNames) {
+		return pgerror.Newf(pgcode.FeatureNotSupported, "cannot use tenant names")
+	}
+
+	if err := rejectIfSystemTenant(tenID, "rename"); err != nil {
+		return err
+	}
+
+	if num, err := p.ExecCfg().InternalExecutor.ExecEx(
+		ctx, "rename-tenant", p.txn, sessiondata.NodeUserSessionDataOverride,
+		`UPDATE system.public.tenants
+SET info =
+crdb_internal.json_to_pb('cockroach.sql.sqlbase.TenantInfo',
+  crdb_internal.pb_to_json('cockroach.sql.sqlbase.TenantInfo', info) ||
+  json_build_object('name', $2))
+WHERE id = $1`, tenID, name); err != nil {
+		if pgerror.GetPGCode(err) == pgcode.UniqueViolation {
+			return pgerror.Newf(pgcode.DuplicateObject, "name %q is already taken", name)
+		}
+		return errors.Wrap(err, "renaming tenant")
```

Comment:
This is more of a general question not really related to this PR, but I think this PR demonstrates the issue:

I noticed we are passing details related to the outer context (via `op` or similar params) in functions above and other places, but why not create a named return param (we use `retErr` in some places`) and wrap any resulting error inside a `defer` with the details from this layer (in this case that we are doing "rename tenant") instead of passing details down the stack? 

The error handling code seems to be inconsistent and is hard to maintain because of the mix of various `op` names being passed to `RequireAdminRole`, `rejectIfSystemTenant`, `ExecEx`, and error wrapping at this layer.